packages/@react-stately/table/src/useTableColumnResizeState.ts

```diff
@@ -0,0 +1,181 @@
+
+import {ColumnProps} from '@react-types/table';
+import {getContentWidth, getDynamicColumnWidths, getMaxWidth, getMinWidth, isStatic, mapToArray, parseStaticWidth} from './utils';
+import {GridNode} from '@react-types/grid';
+import {Key, MutableRefObject, useCallback, useRef, useState} from 'react';
+import {useLayoutEffect} from '@react-aria/utils';
+
+export interface ColumnResizeState<T> {
+  /** A ref whose current value is the state of all the column widths. */
+  columnWidths: MutableRefObject<Map<Key, number>>,
+  /** Setter for the table width. */
+  setTableWidth: (width: number) => void,
+  /** Trigger a resize and recalc. */
+  onColumnResize: (column: GridNode<T>, width: number) => void,
+  /** Callback for when onColumnResize has started. */
+  onColumnResizeStart: () => void,
+    /** Callback for when onColumnResize has ended. */
+  onColumnResizeEnd: (column: GridNode<T>, width: number) => void,
+    /** Getter for column widths. */
+  getColumnWidth(key: Key): number
+}
+
+export interface ColumnResizeStateProps<T> {
+  /** Collection of existing columns. */
+  columns: GridNode<T>[],
+  /** Callback to determine what the default width of a column should be. */
+  getDefaultWidth: (props) => string | number,
+  /** Callback that is invoked during the entirety of the resize event. */
+  onColumnResize?: (affectedColumnWidths: { key: Key, width: number }[]) => void,
+  /** Callback that is invoked when the resize event is ended. */
+  onColumnResizeEnd?: (affectedColumnWidths: { key: Key, width: number }[]) => void
+}
+
+export default function useTableColumnResizeState<T>(props: ColumnResizeStateProps<T>): ColumnResizeState<T> {
+
+  const {columns, getDefaultWidth} = props;
+  const columnsRef = useRef<GridNode<T>[]>();
+  const tableWidth = useRef<number>(null);
+  const isResizing = useRef<boolean>(null);
+  const startResizeContentWidth = useRef<number>();
+
+  const [columnWidths, setColumnWidths] = useState<Map<Key, number>>(new Map(columns.map(col => [col.key, 0])));
+  const columnWidthsRef = useRef<Map<Key, number>>(columnWidths);
+  const [resizedColumns, setResizedColumns] = useState<Set<Key>>(new Set());
+  const resizedColumnsRef = useRef<Set<Key>>(resizedColumns);
+
+  let getRealColumnWidth = useCallback((column: GridNode<T>) => {
```

Comment:
add type for returned value, i thought it was going to always be a number given the name

is this "Real" column width actually "default" or "user assigned" or "starting"?

Real to me sounds like the resolved width after all your maths

Reply:
Haha this is a funny question because it's REALLY hard to name this. It really is the REAL one.

Real is the one that wins after a given order. Like it might be the width that was previously calculated, or if that doesn't exist then it might be user-assigned, it might be a default one assigned by the user, it might be a default one assigned by the developer, or it might be a default width assigned by the code. 

At the end of ALL of this, we have whatever the "real" width is. I'm not sure what else it could be named because it really is the ONE TRUE WIDTH depending on whatever that means at the time it's called, haha.

Reply:
sounds to me like it's `resolveColumnWidth`? or `getResolvedColumnWidth`?

and lets add a quick code comment about order of importantance when resolving that is easier to look at
`// resolve order: prevCalc -> userAssigned -> default user assigned -> default dev assigned -> default in code`

feel free to improve on it, but that looks easier to me than the ternary + nullish coalescing, I know that you can get that from the code, but it's nice to be able to read it in English i think, has a bit more context, i don't have to jump from "props" to "dev assigned"